utf8proc.c

```diff
@@ -310,6 +311,8 @@ static utf8proc_bool grapheme_break_extended(int lbc, int tbc, utf8proc_int32_t
       else
         *state = tbc;
     }
+    else if (*state == UTF8PROC_BOUNDCLASS_EXTEND && tbc == UTF8PROC_BOUNDCLASS_ZWJ)
```

Comment:
Why are we in `UTF8PROC_BOUNDCLASS_EXTEND` at this point rather than `UTF8PROC_BOUNDCLASS_EXTENDED_PICTOGRAPHIC`?  I'm a little confused about how this rule corresponds to GB11.

Reply:
I've replicated the logic that is `utfproc.c` in Julia. 

```julia
julia --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.2 (2020-09-23)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> const utf8proc_propval_t = Int16
Int16

julia> const utf8proc_uint16_t = UInt16
UInt16

julia> struct utf8proc_property_t
         category::utf8proc_propval_t
         combining_class::utf8proc_propval_t
         bidi_class::utf8proc_propval_t
         decomp_type::utf8proc_propval_t
         decomp_seqindex::utf8proc_uint16_t
         casefold_seqindex::utf8proc_uint16_t
         uppercase_seqindex::utf8proc_uint16_t
         lowercase_seqindex::utf8proc_uint16_t
         titlecase_seqindex::utf8proc_uint16_t
         comb_index::utf8proc_uint16_t
         flags::UInt8
         boundclass::UInt8
         # unsigned bidi_mirrored:1;
         # unsigned comp_exclusion:1;
         # unsigned ignorable:1;
         # unsigned control_boundary:1;
         # unsigned charwidth:2;
         # unsigned pad:2;
         # unsigned boundclass:8;
       end

julia> prop(c::Char) = ccall(:utf8proc_get_property, Ref{utf8proc_property_t}, (UInt32, ), c)
prop (generic function with 1 method)

julia> @enum utf8proc_boundclass_t begin
         UTF8PROC_BOUNDCLASS_START              =  0 # /**< Start */
         UTF8PROC_BOUNDCLASS_OTHER              =  1 # /**< Other */
         UTF8PROC_BOUNDCLASS_CR                 =  2 # /**< Cr */
         UTF8PROC_BOUNDCLASS_LF                 =  3 # /**< Lf */
         UTF8PROC_BOUNDCLASS_CONTROL            =  4 # /**< Control */
         UTF8PROC_BOUNDCLASS_EXTEND             =  5 # /**< Extend */
         UTF8PROC_BOUNDCLASS_L                  =  6 # /**< L */
         UTF8PROC_BOUNDCLASS_V                  =  7 # /**< V */
         UTF8PROC_BOUNDCLASS_T                  =  8 # /**< T */
         UTF8PROC_BOUNDCLASS_LV                 =  9 # /**< Lv */
         UTF8PROC_BOUNDCLASS_LVT                = 10 # /**< Lvt */
         UTF8PROC_BOUNDCLASS_REGIONAL_INDICATOR = 11 # /**< Regional indicator */
         UTF8PROC_BOUNDCLASS_SPACINGMARK        = 12 # /**< Spacingmark */
         UTF8PROC_BOUNDCLASS_PREPEND            = 13 # /**< Prepend */
         UTF8PROC_BOUNDCLASS_ZWJ                = 14 # /**< Zero Width Joiner */

         # the following are no longer used in Unicode 11, but we keep
         # the constants here for backward compatibility
         UTF8PROC_BOUNDCLASS_E_BASE             = 15 # /**< Emoji Base */
         UTF8PROC_BOUNDCLASS_E_MODIFIER         = 16 # /**< Emoji Modifier */
         UTF8PROC_BOUNDCLASS_GLUE_AFTER_ZWJ     = 17 # /**< Glue_After_ZWJ */
         UTF8PROC_BOUNDCLASS_E_BASE_GAZ         = 18 # /**< E_BASE + GLUE_AFTER_ZJW */

         # the Extended_Pictographic property is used in the Unicode 11
         # grapheme-boundary rules, so we store it in the boundclass field
         UTF8PROC_BOUNDCLASS_EXTENDED_PICTOGRAPHIC = 19
         UTF8PROC_BOUNDCLASS_E_ZWG = 20 # /* UTF8PROC_BOUNDCLASS_EXTENDED_PICTOGRAPHIC + ZWJ */
       end

julia> function boundclass(c::Char)
         r = utf8proc_boundclass_t(prop(c).boundclass)
         return r
       end
boundclass (generic function with 1 method)
```

If we take the following unicode string, can call the boundclass function on it, we get the following:

```julia
julia> unicode_s = "🤦🏼‍♂️"
"🤦🏼\u200d♂️"

julia> for c in unicode_s
       @show boundclass(c)
       end
boundclass(c) = UTF8PROC_BOUNDCLASS_EXTENDED_PICTOGRAPHIC
boundclass(c) = UTF8PROC_BOUNDCLASS_EXTEND
boundclass(c) = UTF8PROC_BOUNDCLASS_ZWJ
boundclass(c) = UTF8PROC_BOUNDCLASS_EXTENDED_PICTOGRAPHIC
boundclass(c) = UTF8PROC_BOUNDCLASS_EXTEND
```

I've not read the spec enough to know whether the implementation here is incorrect and if those codepoints should correspond to `EXTENDED_PICTOGRAPHIC` or `EXTEND`.

Reply:
The codepoints are correct, but why does the `state` not stay in `UTF8PROC_BOUNDCLASS_EXTENDED_PICTOGRAPHIC` since that is the first character?

Reply:
This is because the first "state" is always `UTF8PROC_BOUNDCLASS_START`. And in this case, it is `UTF8PROC_BOUNDCLASS_START` instead of `UTF8PROC_BOUNDCLASS_EXTENDED_PICTOGRAPHIC`. It then goes to `UTF8PROC_BOUNDCLASS_EXTENDED` instead.

Reply:
This is also perhaps why it works as expected for strings where an ascii character is prefixed with the emoji, and we only encounter this issue when the emoji is at the very start of the string.

Reply:
Shouldn't the state change to `UTF8PROC_BOUNDCLASS_EXTENDED_PICTOGRAPHIC` after the first (emoji) codepoint?